### PR TITLE
New rule: preserve-di-assignment-case

### PR DESCRIPTION
**Pull request by TomMcQuarrieOTH**
Shiny new rule: preserve-di-assignment-case

### CODE REVIEW

**Review by lee-borsboom — APPROVED** [review]
Looks good to me, with the caveat that I know very little about this stuff.

File: tests/lib/rules/preserve-di-assignment-case.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    constructor(MyService, MyService2) {
+        this.MyService = MyService;
+        this.MyService2 = MyService2;
+    }
+}`],
+    invalid: [
+        {
+            code: `
+class MyClass {
+    constructor(MyService, MyService2, MyService3) {
```
Comment (kylezeeuwen) [inline tests/lib/rules/preserve-di-assignment-case.js]: should these by myService not nyService ?

**Comment by kylezeeuwen** [comment]
Only concern is whether we should allow someone to completely rename something. My initial thoughts on the rule was that if the author has completely renamed the service upon assignment then thats ok, the only case I was trying to catch was when they changed _only_ the capitalisation ...

that said Im happy giving this a try and seeing what comes of it. How many exceptions does this pick up? I can think of times when I have seen people take (CONFIG) -> { this.B = CONFIG.B } . Would this trigger a warning/error ? If so I think we need to relax the rule

File: lib/utils/utils.js [hunk 2]
```diff
@@ -12,6 +12,21 @@ module.exports = {
     return !methods || methods.length === 0;
   },
 
+  getAssignmentsToThis: function(node) {
+    return _.filter(node.value.body.body, (item) => {
```
Comment (paulbevis) [inline lib/utils/utils.js:16]: I guess you could remove the if statement if you wanted to make it more concise...

File: lib/rules/preserve-di-assignment-case.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      description: "Should preserve the import name and case of injected items assigned to this",
+      category: "Possible Errors",
+      recommended: true
+    },
+    schema: []
+  },
+  create: (context) => {
+    return {
+      'ClassDeclaration': (node) => {
+        // class has no methods/constructor, so nothing to inject...
```
Comment (paulbevis) [inline lib/rules/preserve-di-assignment-case.js]: Being really picky ;) Is that a cut 'n' paste in the comments regarding 'nothing to inject'...

File: tests/lib/rules/preserve-di-assignment-case.js (hunk 1 above)
Comment (TomMcQuarrieOTH) [reply tests/lib/rules/preserve-di-assignment-case.js]: I intended on having 1 as a typo but yep this needs to be more explicit

File: lib/utils/utils.js (hunk 2 above)
Comment (TomMcQuarrieOTH) [reply lib/utils/utils.js:16]: Ah, so you mean just do
`return item.type === 'ExpressionStatement' && item.expression.type === 'AssignmentExpression' && ...`

Yep a little more concise, thanks!

**Review by paulbevis — APPROVED** [review]

**Pull request merged**